### PR TITLE
Sending show plan graph to ADS on Result Set updated event

### PR DESCRIPTION
**Pull request by aasimkhan30**
This PR does 2 following things 
1. Sending show plan graph on result set updated event to the frontend.
1. I have converted the showplan graph to simpler object to avoid circular references.

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings [hunk 1]
```diff
@@ -2213,6 +2213,16 @@ SizeInMegaBytesFormat = {0} MB
 SizeInGigaBytesFormat = {0} GB
 ;Size in TeraBytes format
 SizeInTeraBytesFormat = {0} TB
+; display string for the operator cost property - 0.###### - is the float number format specifier
+OperatorDisplayCost(double cost, int percentage) = {0:0.#######} ({1}%)
+;This node name will be visible in Active Accesibility
+NodeDisplayPropertiesName2(string title, long nodeId) = {0}[{1}]
+;This node name will be visible in Active Accesibility
+NodeDisplayPropertiesName1(long nodeId) = [{0}]
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings:2221]: These don't seem like things that actually need to be localized - there's no text to localize here

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+using System.Collections.Generic;
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs:1]: copyright

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs [hunk 3]
```diff
@@ -55,6 +57,186 @@ public int GroupIndex
...
+                }
+
+                return secondLine == null || secondLine.Length == 0
+                    ? firstLine
+                    : String.Format(CultureInfo.CurrentCulture, "{0}\n{1}", firstLine, secondLine);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets node name (visible through AA)
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:128]: What's AA? And `Name2` isn't very descriptive, is there not something better you could use?

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs [hunk 4]
```diff
@@ -55,6 +57,186 @@ public int GroupIndex
...
+        /// </summary>
+        public string Name2
+        {
+            get
+            {
+                // Make AA names compatible with the previos implementation
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:134]: What does this mean? What previous implementation? Remember people reading this code have no context for where this came from originally so explaining why this stuff is actually necessary is better.

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs [hunk 5]
```diff
@@ -55,6 +57,186 @@ public int GroupIndex
...
+                }
+                else
+                {
+                    return SR.NodeDisplayPropertiesName2(name, this.ID);
+                }
+            }
+
+            set
+            {
+                ;
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:150]: Is this needed?

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs [hunk 6]
```diff
@@ -55,6 +57,186 @@ public int GroupIndex
...
+        /// Gets the value that indicates whether the Node has critical warnings.
+        /// </summary>
+        private bool HasCriticalWarnings
+        {
+            get
+            {
+                if (this["Warnings"] != null)
+                {
+                    ExpandableObjectWrapper wrapper = this["Warnings"] as ExpandableObjectWrapper;
+                    if (wrapper["NoJoinPredicate"] != null)
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:198]: Why this specific property?

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/ShowPlan/ShowPlanTests.cs [hunk 7]
```diff
@@ -16,17 +17,15 @@ namespace Microsoft.SqlTools.ServiceLayer.UnitTests.ShowPlan
...
             StreamReader reader = new StreamReader(scriptStream);
             string text = reader.ReadToEnd();
-            var showPlanGraphs = ShowPlanGraph.ParseShowPlanXML(text, ShowPlanType.Actual);
-            Assert.AreEqual(1, showPlanGraphs.Length, "Single show plan graph not generated from the test xml file");
-            var testShowPlanGraph = showPlanGraphs[0];
-            Assert.NotNull(testShowPlanGraph, "graph should not be null");
-            Assert.NotNull(testShowPlanGraph.Root, "graph should have a root");
+            var showPlanGraphs = ShowPlanGraphUtils.CreateShowPlanGraph(text);
+            Assert.NotNull(showPlanGraphs, "graph should not be null");
+            Assert.NotNull(showPlanGraphs.Root, "graph should have a root");
```
Comment (Charles-Gagnon) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/ShowPlan/ShowPlanTests.cs:28]: Why did you remove these other checks? Are there no other checks you can do?

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs [hunk 8]
```diff
@@ -893,13 +894,18 @@ public async Task<ResultSetSubset> InterServiceResultSubset(SubsetParams subsetP
...
+                {
+                    var xmlString = r.GetRow(0)[0].DisplayValue;
+                    graph = ShowPlanGraphUtils.CreateShowPlanGraph(xmlString);
+                }
                 ResultSetUpdatedEventParams eventParams = new ResultSetUpdatedEventParams
                 {
                     ResultSetSummary = r.Summary,
-                    OwnerUri = ownerUri
+                    OwnerUri = ownerUri,
+                    ShowPlanGraph = graph
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:907]: I'm concerned about sending over a huge showplan graph - have you tested if there's any perf issues with those?

**Review by Charles-Gagnon — COMMENTED** [review]
There seems to be a lot more than just sending the showplan graph, in the future please either make sure the title/description is accurate or better yet split into multiple PRs.

File: src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings (hunk 1 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings:2221]: I got these strings from SSMS. Should I move them somewhere else? I see that they have added some basic formated texts in .strings file.

Comment (Charles-Gagnon) [reply src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings:2221]: I would just embed them directly into the code using string interpolation or string.Format

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs (hunk 4 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:134]: I got these AA comments from SSMS. Will remove this prop as it is not used currently.

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs (hunk 6 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:171]: Got this piece of code from SSMS. I haven't really investigated the logic of this property

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs (hunk 8 above)
Comment (aasimkhan30) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:907]: Since we are generating the show plan graph only for the last result set update, I don't think it will be a perf issue for the users. Also, most of the show plans graph are not excessively big.

Comment (Charles-Gagnon) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:907]: When it's generated doesn't really matter if the size is too large.

How are you coming to your statement "most of the show plan graph are not that excessively big"? I'm not sure that's something we can take for granted. I would suggest talking to the QP experts in that area to get a sense for whether this is something we should be more careful of. And at the very least you should be doing testing to see if there's a limit for when perf starts to become a noticeable issue.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs [hunk 9]
```diff
@@ -14,6 +14,7 @@
 using Microsoft.SqlTools.ServiceLayer.QueryExecution.Contracts.ExecuteRequests;
 using Microsoft.SqlTools.ServiceLayer.QueryExecution.DataStorage;
 using Microsoft.SqlTools.ServiceLayer.SqlContext;
+using Microsoft.SqlTools.ServiceLayer.ShowPlan;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:17]: sort order of usings looks incorrect.  Also, I see one out of order below.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs [hunk 10]
```diff
@@ -893,13 +894,18 @@ public async Task<ResultSetSubset> InterServiceResultSubset(SubsetParams subsetP
             // Setup the ResultSet updated callback
             ResultSet.ResultSetAsyncEventHandler resultUpdatedCallback = async r =>
             {
+                ExecutionPlanGraph graph = null;
+                if(r.Summary.Complete && r.Summary.SpecialAction.ExpectYukonXMLShowPlan == true)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:898]: nit: add space between 'f' and '('.  Also comparing `ExpectYukonXMLShowPlan` with `true` is somewhat redundant since that is default test in if condition.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs [hunk 11]
```diff
@@ -893,13 +894,18 @@ public async Task<ResultSetSubset> InterServiceResultSubset(SubsetParams subsetP
...
+                {
+                    var xmlString = r.GetRow(0)[0].DisplayValue;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:900]: I guess we always know that we can safely reference this element with `(0)[0]`.  i.e. no needs to bounds-check?

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs (hunk 8 above)
Comment (kburtram) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs:907]: we should look at the perf of this scenario to understand it.  but we need to transmit the graph data to the client to render it at some point.  I guess maybe the XML is more compact than the processed JSON format. Though I suspect compared to the size of other things we pass around (e.g. query resultsets).  I suspect the delta between the XML format and JSON format isn't that impactful.

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+{
+    /// <summary>
+    /// Execution plan graph object that is sent over JSON RPC
+    /// </summary>
+    public class ExecutionPlanGraph
+    {
+        /// <summary>
+        /// Root of the execution plan tree
+        /// </summary>
+        public ExecutionPlanNode Root {get; set;}
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs:13]: inconsistent spacing around the '{' and 'g' and ';' and '}' compared to other class.

**Review by kburtram — APPROVED** [review]
Looks reasonable for a work in progress.  Several items that can be cleaned up.  But mostly fine for this PR.

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs (hunk 2 above)
Comment (kburtram) [reply src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs:1]: do need to push more changes to resolve this?

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        /// </summary>
+        public ExecutionPlanNode Root { get; set; }
+        /// <summary>
+        /// Underlying query for the execution plan graph
+        /// </summary>
+        public string Query { get; set; }
+        /// <summary>
+        ///  Execution plan graph error
+        /// </summary>
+        public string error { get; set; }
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/Contracts/ExecutionPlanGraph.cs:21]: `error` should be capitalized

File: src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs [hunk 14]
```diff
@@ -55,6 +57,159 @@ public int GroupIndex
...
+                // The first line is the Physical name and the physical kind in parenthesis
+                // The second line should contains either Object value or LogicalOp name.
+                // The second line should not show the same content as the first line.
+
+                string firstLine = this["PhysicalOp"] as string;
+                if (firstLine == null)
+                {
+                    if (this.Operation == null)
+                    {
+                        return String.Empty;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/ShowPlan/ShowPlanGraph/Node.cs:82]: nit: `string.Empty` is preferred to `String.Empty` per .Net Framework Coding Guidelines.  Basically primitive types should be preferred to BCL classes when both are available.  Probably not worth rewriting the code, more FYI for next time.

**Review by kburtram — APPROVED** [review]
Looks mostly fine for hooking up this initial event.  I think there is a missing copyright and some other pending comments.  Though overall once major issues are addressed, I push this and take remaining issues for follow-up PRs as we iterate on the feature.
